pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,235 @@
+            self._bagua_kwargs["q_adam_optimizer"] = self.optimizers[0]
+
+        algorithm = Algorithm.init(self._bagua_algorithm, **self._bagua_kwargs)
+        return BaguaDistributedDataParallel(
+            module=model,
+            optimizers=self.optimizers,
+            algorithm=algorithm,
+            gradient_as_bucket_view=self._bagua_do_flatten,
+        )
+
+    @property
+    def lightning_module(self) -> "pl.LightningModule":
+        model = self._model
+        if isinstance(model, BaguaDistributedDataParallel):
+            model = model.module
+        return unwrap_lightning_module(model)  # type: ignore[arg-type]
+
+    @classmethod
+    def register_plugins(cls, plugin_registry: Dict) -> None:
+        plugin_registry.register("bagua", cls, description="Default Bagua Plugin")
+
+    def teardown(self) -> None:
+        # abort the background communication for async algorithm
+        if self.lightning_module.trainer.training and self._bagua_algorithm == "async":
+            self.model.bagua_algorithm.abort(self.model)  # type: ignore
+
+        if self.on_gpu:
+            # GPU teardown
+            self.lightning_module.cpu()
+            # clean up memory
+            torch.cuda.empty_cache()
+
+    def barrier(self, *args, **kwargs) -> None:  # type: ignore[no-untyped-def]
+        if is_initialized():
+            barrier()
+
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        return broadcast_object(obj, src)
+
+    def reduce(
+        self, tensor: torch.Tensor, group: Optional[Any] = None, reduce_op: Optional[Union[ReduceOp, str]] = "mean"
+    ) -> torch.Tensor:
+        """Reduces a tensor from several distributed processes to one aggregated tensor.
+
+        Args:
+            tensor: The tensor to sync and reduce.
+            group: The process group to gather results from. Defaults to all processes (world).
+            reduce_op: The reduction operation. Defaults to 'mean'.
+                Can also be a string 'sum' or ReduceOp.
+
+        Return:
+            The reduced value, except when the input was not a tensor the output remains is unchanged.
+        """
+        if not isinstance(tensor, torch.Tensor):
+            return tensor
+        if group is not None:
+            raise ValueError("Bagua does not support allreduce using a subcommunicator at this time. Unset `group`.")
+
+        if reduce_op is None:
+            op = BaguaReduceOp.AVG
+        elif isinstance(reduce_op, str):
+            if reduce_op.lower() in ("avg", "mean"):
+                op = BaguaReduceOp.AVG
+            elif reduce_op.lower() == "sum":
+                op = BaguaReduceOp.SUM
+            else:
+                raise ValueError(f"unrecognized `reduce_op`: {reduce_op}")
+        elif isinstance(reduce_op, ReduceOp):
+            op = from_torch_reduce_op(reduce_op)
+            if op is None:
+                raise ValueError(f"unrecognized `reduce_op`: {reduce_op}")
```

Comment:
```suggestion
                raise ValueError(f"Unrecognized `reduce_op`: {reduce_op}")
        elif isinstance(reduce_op, ReduceOp):
            op = from_torch_reduce_op(reduce_op)
            if op is None:
                raise ValueError(f"Unrecognized `reduce_op`: {reduce_op}")
```
Can we add better error msg? something like `BaguaStrategy doesn't support reduce_op: {reduce_op}`, just to let users know where the error is actually coming from.

Reply:
done